File: examples/python/compatibility_example.py
#!/usr/bin/env python3
"""
Hydrogen Automatic ASCOM/INDI Compatibility Example

This example demonstrates how to use the automatic ASCOM/INDI compatibility
system from Python, enabling seamless protocol integration for devices.
"""

import sys
import time
import os
import threading
from typing import Optional

# Add the build directory to Python path
sys.path.append(os.path.join(os.path.dirname(__file__), '../../build'))

try:
    import pyhydrogen as hydrogen
except ImportError:
    print("Error: pyhydrogen module not found. Make sure to build the Python bindings first.")
    print("Run: cmake --build build --target pyhydrogen")
    sys.exit(1)

class CompatibilityDemo:
    """Demonstrates automatic ASCOM/INDI compatibility features"""
    
    def __init__(self):
        self.devices = {}
        self.running = False
        
    def initialize_system(self):
        """Initialize the automatic compatibility system"""
        print("🚀 Initializing Hydrogen Automatic ASCOM/INDI Compatibility System")
        
        # Set logging level
        hydrogen.set_log_level("info")
        
        # Initialize the compatibility system
        hydrogen.init_compatibility_system(
            enable_auto_discovery=True,
            enable_ascom=True,
            enable_indi=True,
            indi_base_port=7624
        )
        
        print("✅ Compatibility system initialized")
        
    def create_compatible_devices(self):
        """Create devices with automatic ASCOM/INDI compatibility"""
        print("\n📷 Creating devices with automatic compatibility...")
        
        try:
            # Create camera with automatic compatibility
            self.devices['camera'] = hydrogen.create_compatible_camera(
                "demo_camera", "ZWO", "ASI294MM Pro"
            )
            print("  ✅ Camera created with ASCOM/INDI compatibility")
            
            # Create telescope with automatic compatibility
            self.devices['telescope'] = hydrogen.create_compatible_telescope(
                "demo_telescope", "Celestron", "CGX-L"
            )
            print("  ✅ Telescope created with ASCOM/INDI compatibility")
            
            # Create focuser with automatic compatibility
            self.devices['focuser'] = hydrogen.create_compatible_focuser(
                "demo_focuser", "ZWO", "EAF"
            )
            print("  ✅ Focuser created with ASCOM/INDI compatibility")
            
            # Create dome with automatic compatibility
            self.devices['dome'] = hydrogen.create_compatible_dome(
                "demo_dome", "Generic", "Observatory Dome"
            )
            print("  ✅ Dome created with ASCOM/INDI compatibility")
            
            # Create observing conditions with automatic compatibility
            self.devices['weather'] = hydrogen.create_compatible_observing_conditions(
                "demo_weather", "Generic", "Weather Station"
            )
            print("  ✅ Weather station created with ASCOM/INDI compatibility")
            
        except Exception as e:
            print(f"  ❌ Error creating devices: {e}")
            return False
            
        return True
        
    def demonstrate_device_operations(self):
        """Demonstrate basic device operations"""
        print("\n🔧 Demonstrating device operations...")
        
        try:
            # Camera operations
            camera = self.devices.get('camera')
            if camera:
                print("  📷 Camera operations:")
                camera.set_property("coolerOn", True)
                camera.set_property("temperature", -10.0)
                print(f"    - Cooler enabled, target temperature: {camera.get_property('temperature')}°C")
                
                # Start a short exposure
                camera.start_exposure(1.0)  # 1 second exposure
                print("    - Started 1-second exposure")
                
            # Telescope operations
            telescope = self.devices.get('telescope')
            if telescope:
                print("  🔭 Telescope operations:")
                telescope.set_property("tracking", True)
                print("    - Tracking enabled")
                
                # Set target coordinates (example: M42)
                telescope.set_property("targetRA", 5.59)  # 5h 35m
                telescope.set_property("targetDec", -5.39)  # -5° 23'
                print("    - Target set to M42 (Orion Nebula)")
                
            # Focuser operations
            focuser = self.devices.get('focuser')
            if focuser:
                print("  🎯 Focuser operations:")
                current_pos = focuser.get_property("position")
                new_pos = current_pos + 100
                focuser.set_property("targetPosition", new_pos)
                print(f"    - Moving from {current_pos} to {new_pos} steps")
                
            # Dome operations
            dome = self.devices.get('dome')
            if dome:
                print("  🏠 Dome operations:")
                dome.open_shutter()
                dome.slew_to_azimuth(180.0)  # Point south
                print("    - Opening shutter and slewing to south")
                
            # Weather station operations
            weather = self.devices.get('weather')
            if weather:
                print("  🌤️ Weather station operations:")
                weather.refresh()
                temp = weather.get_temperature()
                humidity = weather.get_humidity()
                print(f"    - Temperature: {temp}°C, Humidity: {humidity}%")
                
        except Exception as e:
            print(f"  ❌ Error in device operations: {e}")
            
    def demonstrate_protocol_transparency(self):
        """Demonstrate transparent protocol access"""
        print("\n🔄 Demonstrating protocol transparency...")
        
        try:
            camera = self.devices.get('camera')
            if camera:
                print("  📷 Camera protocol transparency:")
                
                # The same property accessed through different protocols
                # Note: In the actual implementation, you would access the bridge
                # For this demo, we'll show the concept
                
                # Internal protocol (direct device access)
                internal_temp = camera.get_property("temperature")
                print(f"    - Internal protocol temperature: {internal_temp}°C")
                
                # ASCOM protocol (would be CCDTemperature)
                print("    - ASCOM protocol: Property automatically mapped to CCDTemperature")
                
                # INDI protocol (would be CCD_TEMPERATURE)
                print("    - INDI protocol: Property automatically mapped to CCD_TEMPERATURE")
                
                print("    - All protocols stay synchronized automatically! 🎯")
                
        except Exception as e:
            print(f"  ❌ Error demonstrating protocol transparency: {e}")
            
    def show_system_statistics(self):
        """Show compatibility system statistics"""
        print("\n📊 System Statistics:")
        
        try:
            stats = hydrogen.get_compatibility_statistics()
            
            print(f"  📈 Total devices: {stats.total_devices}")
            print(f"  🔌 ASCOM-enabled devices: {stats.ascom_devices}")
            print(f"  🌐 INDI-enabled devices: {stats.indi_devices}")
            print(f"  ⏱️ System uptime: {stats.uptime.total_seconds():.1f} seconds")
            
            if stats.device_type_count:
                print("  📋 Device types:")
                for device_type, count in stats.device_type_count.items():
                    print(f"    - {device_type}: {count}")
                    
        except Exception as e:
            print(f"  ❌ Error getting statistics: {e}")
            
    def simulate_ascom_client(self):
        """Simulate an ASCOM client connecting to devices"""
        print("\n🔌 Simulating ASCOM client connections...")
        
        print("  🎯 ASCOM clients can now connect to:")
        for device_name, device in self.devices.items():
            device_type = device.get_device_type()
            device_id = device.get_device_id()
            print(f"    - {device_type} ({device_id}) via ASCOM COM interface")
            
        print("  📝 Example ASCOM client code (VB.NET/C#):")
        print("    Dim camera As Object = CreateObject(\"Hydrogen.Camera.demo_camera\")")
        print("    camera.Connected = True")
        print("    camera.StartExposure(5.0, True)")
        
    def simulate_indi_client(self):
        """Simulate an INDI client connecting to devices"""
        print("\n🌐 Simulating INDI client connections...")
        
        print("  🎯 INDI clients can now connect to:")
        for device_name, device in self.devices.items():
            device_type = device.get_device_type()
            device_id = device.get_device_id()
            print(f"    - {device_type} ({device_id}) via INDI XML protocol on port 7624+")
            
        print("  📝 Example INDI client code (Python):")
        print("    import PyIndi")
        print("    client = PyIndi.BaseClient()")
        print("    client.setServer(\"localhost\", 7624)")
        print("    client.connectServer()")
        
    def cleanup(self):
        """Clean up resources"""
        print("\n🧹 Cleaning up...")
        
        try:
            # Stop all devices
            for device_name, device in self.devices.items():
                try:
                    device.stop()
                    print(f"  ✅ Stopped {device_name}")
                except Exception as e:
                    print(f"  ⚠️ Error stopping {device_name}: {e}")
                    
            # Shutdown compatibility system
            hydrogen.shutdown_compatibility_system()
            print("  ✅ Compatibility system shutdown")
            
        except Exception as e:
            print(f"  ❌ Error during cleanup: {e}")
            
    def run_demo(self):
        """Run the complete demonstration"""
        print("=" * 60)
        print("🌟 Hydrogen Automatic ASCOM/INDI Compatibility Demo")
        print("=" * 60)
        
        try:
            # Initialize system
            self.initialize_system()
            
            # Create devices
            if not self.create_compatible_devices():
                return
                
            # Wait a moment for initialization
            time.sleep(1)
            
            # Demonstrate operations
            self.demonstrate_device_operations()
            
            # Show protocol transparency
            self.demonstrate_protocol_transparency()
            
            # Show statistics
            self.show_system_statistics()
            
            # Simulate client connections
            self.simulate_ascom_client()
            self.simulate_indi_client()
            
            print("\n🎉 Demo completed successfully!")
            print("\n💡 Key Benefits:")
            print("  ✅ Zero code changes required for existing devices")
            print("  ✅ Automatic ASCOM/INDI protocol support")
            print("  ✅ Transparent property synchronization")
            print("  ✅ Universal client compatibility")
            print("  ✅ Real-time protocol bridging")
            
        except KeyboardInterrupt:
            print("\n⏹️ Demo interrupted by user")
        except Exception as e:
            print(f"\n❌ Demo error: {e}")
        finally:
            self.cleanup()

def main():
    """Main function"""
    demo = CompatibilityDemo()
    demo.run_demo()

if __name__ == "__main__":
    main()

File: src/device/interfaces/compatibility_example.cpp
#include "integration_manager.h"
#include "../camera.h"
#include "../telescope.h"
#include "../focuser.h"
#include <iostream>
#include <thread>
#include <chrono>

using namespace hydrogen::device;
using namespace hydrogen::device::interfaces;

/**
 * @brief Comprehensive example demonstrating automatic ASCOM/INDI compatibility
 * 
 * This example shows how the automatic compatibility system works without
 * requiring any changes to existing device implementations.
 */
class CompatibilityExample {
public:
    void runExample() {
        std::cout << "=== Hydrogen Automatic ASCOM/INDI Compatibility Example ===" << std::endl;
        
        // Initialize the integration manager
        initializeIntegrationManager();
        
        // Create and register devices
        createAndRegisterDevices();
        
        // Demonstrate transparent protocol access
        demonstrateProtocolAccess();
        
        // Demonstrate ASCOM client simulation
        demonstrateASCOMClient();
        
        // Demonstrate INDI client simulation
        demonstrateINDIClient();
        
        // Show statistics
        showStatistics();
        
        // Cleanup
        cleanup();
        
        std::cout << "=== Example completed ===" << std::endl;
    }

private:
    integration::AutomaticIntegrationManager* manager_;
    std::shared_ptr<Camera> camera_;
    std::shared_ptr<Telescope> telescope_;
    std::shared_ptr<Focuser> focuser_;
    
    void initializeIntegrationManager() {
        std::cout << "\n1. Initializing Integration Manager..." << std::endl;
        
        manager_ = &integration::AutomaticIntegrationManager::getInstance();
        
        // Configure integration
        integration::IntegrationConfiguration config;
        config.autoDiscovery = true;
        config.autoRegistration = true;
        config.enableASCOM = true;
        config.enableINDI = true;
        config.deviceNamePrefix = "HydrogenExample_";
        
        // Set device-specific configurations
        bridge::BridgeConfiguration cameraConfig("HydrogenExample_Camera", "Example Camera Device");
        cameraConfig.customProperties["manufacturer"] = "Hydrogen";
        cameraConfig.customProperties["model"] = "ExampleCam";
        config.deviceConfigs["camera1"] = cameraConfig;
        
        bridge::BridgeConfiguration telescopeConfig("HydrogenExample_Telescope", "Example Telescope Device");
        telescopeConfig.customProperties["manufacturer"] = "Hydrogen";
        telescopeConfig.customProperties["model"] = "ExampleScope";
        config.deviceConfigs["telescope1"] = telescopeConfig;
        
        bridge::BridgeConfiguration focuserConfig("HydrogenExample_Focuser", "Example Focuser Device");
        focuserConfig.customProperties["manufacturer"] = "Hydrogen";
        focuserConfig.customProperties["model"] = "ExampleFocus";
        config.deviceConfigs["focuser1"] = focuserConfig;
        
        manager_->initialize(config);
        manager_->start();
        
        // Add discovery callbacks
        manager_->addDeviceDiscoveryCallback([](const std::string& deviceId, std::shared_ptr<core::IDevice> device) {
            std::cout << "  Device discovered: " << deviceId << std::endl;
        });
        
        manager_->addDeviceRemovalCallback([](const std::string& deviceId) {
            std::cout << "  Device removed: " << deviceId << std::endl;
        });
        
        std::cout << "  Integration manager initialized and started" << std::endl;
    }
    
    void createAndRegisterDevices() {
        std::cout << "\n2. Creating and Registering Devices..." << std::endl;
        
        // Create devices using existing implementations
        camera_ = std::make_shared<Camera>("camera1", "Hydrogen", "ExampleCam");
        telescope_ = std::make_shared<Telescope>("telescope1", "Hydrogen", "ExampleScope");
        focuser_ = std::make_shared<Focuser>("focuser1", "Hydrogen", "ExampleFocus");
        
        // Initialize devices
        camera_->initializeDevice();
        camera_->startDevice();
        
        telescope_->initializeDevice();
        telescope_->startDevice();
        
        focuser_->initializeDevice();
        focuser_->startDevice();
        
        // Register devices - this automatically creates ASCOM/INDI bridges
        REGISTER_DEVICE_AUTO("camera1", camera_);
        REGISTER_DEVICE_AUTO("telescope1", telescope_);
        REGISTER_DEVICE_AUTO("focuser1", focuser_);
        
        std::cout << "  Devices registered with automatic ASCOM/INDI compatibility" << std::endl;
        
        // Wait for registration to complete
        std::this_thread::sleep_for(std::chrono::milliseconds(500));
    }
    
    void demonstrateProtocolAccess() {
        std::cout << "\n3. Demonstrating Transparent Protocol Access..." << std::endl;
        
        try {
            // Access camera properties through different protocols
            std::cout << "  Camera Properties:" << std::endl;
            
            // Internal protocol access
            double internalTemp = GET_DEVICE_PROPERTY_AUTO(Camera, "camera1", "temperature", double, bridge::ProtocolType::INTERNAL);
            std::cout << "    Internal temperature: " << internalTemp << "°C" << std::endl;
            
            // ASCOM protocol access (same property, different protocol)
            double ascomTemp = GET_DEVICE_PROPERTY_AUTO(Camera, "camera1", "CCDTemperature", double, bridge::ProtocolType::ASCOM);
            std::cout << "    ASCOM temperature: " << ascomTemp << "°C" << std::endl;
            
            // INDI protocol access
            double indiTemp = GET_DEVICE_PROPERTY_AUTO(Camera, "camera1", "CCD_TEMPERATURE", double, bridge::ProtocolType::INDI);
            std::cout << "    INDI temperature: " << indiTemp << "°C" << std::endl;
            
            // Set properties through different protocols
            SET_DEVICE_PROPERTY_AUTO(Camera, "camera1", "coolerOn", true, bridge::ProtocolType::INTERNAL);
            std::cout << "    Cooler turned on via internal protocol" << std::endl;
            
            SET_DEVICE_PROPERTY_AUTO(Camera, "camera1", "CoolerOn", true, bridge::ProtocolType::ASCOM);
            std::cout << "    Cooler confirmed via ASCOM protocol" << std::endl;
            
            // Telescope properties
            std::cout << "  Telescope Properties:" << std::endl;
            
            double ra = GET_DEVICE_PROPERTY_AUTO(Telescope, "telescope1", "rightAscension", double, bridge::ProtocolType::INTERNAL);
            double dec = GET_DEVICE_PROPERTY_AUTO(Telescope, "telescope1", "declination", double, bridge::ProtocolType::INTERNAL);
            std::cout << "    Current position: RA=" << ra << "h, Dec=" << dec << "°" << std::endl;
            
            // Focuser properties
            std::cout << "  Focuser Properties:" << std::endl;
            
            int position = GET_DEVICE_PROPERTY_AUTO(Focuser, "focuser1", "position", int, bridge::ProtocolType::INTERNAL);
            std::cout << "    Current position: " << position << " steps" << std::endl;
            
        } catch (const std::exception& e) {
            std::cout << "    Error: " << e.what() << std::endl;
        }
    }
    
    void demonstrateASCOMClient() {
        std::cout << "\n4. Simulating ASCOM Client Access..." << std::endl;
        
        try {
            // Simulate ASCOM client connecting to camera
            std::cout << "  ASCOM Client connecting to camera..." << std::endl;
            
            // Get ASCOM standard properties
            auto name = GET_DEVICE_PROPERTY_AUTO(Camera, "camera1", "Name", std::string, bridge::ProtocolType::ASCOM);
            auto description = GET_DEVICE_PROPERTY_AUTO(Camera, "camera1", "Description", std::string, bridge::ProtocolType::ASCOM);
            auto driverVersion = GET_DEVICE_PROPERTY_AUTO(Camera, "camera1", "DriverVersion", std::string, bridge::ProtocolType::ASCOM);
            
            std::cout << "    Device Name: " << name << std::endl;
            std::cout << "    Description: " << description << std::endl;
            std::cout << "    Driver Version: " << driverVersion << std::endl;
            
            // Connect to device
            SET_DEVICE_PROPERTY_AUTO(Camera, "camera1", "Connected", true, bridge::ProtocolType::ASCOM);
            bool connected = GET_DEVICE_PROPERTY_AUTO(Camera, "camera1", "Connected", bool, bridge::ProtocolType::ASCOM);
            std::cout << "    Connected: " << (connected ? "Yes" : "No") << std::endl;
            
            // Start exposure using ASCOM method
            std::cout << "    Starting 5-second exposure..." << std::endl;
            INVOKE_DEVICE_METHOD_AUTO(Camera, void, "camera1", "StartExposure", bridge::ProtocolType::ASCOM, 5.0, true);
            
            // Check exposure status
            bool imageReady = GET_DEVICE_PROPERTY_AUTO(Camera, "camera1", "ImageReady", bool, bridge::ProtocolType::ASCOM);
            std::cout << "    Image Ready: " << (imageReady ? "Yes" : "No") << std::endl;
            
        } catch (const std::exception& e) {
            std::cout << "    Error: " << e.what() << std::endl;
        }
    }
    
    void demonstrateINDIClient() {
        std::cout << "\n5. Simulating INDI Client Access..." << std::endl;
        
        try {
            // Simulate INDI client connecting to telescope
            std::cout << "  INDI Client connecting to telescope..." << std::endl;
            
            // Get INDI properties
            double currentRA = GET_DEVICE_PROPERTY_AUTO(Telescope, "telescope1", "EQUATORIAL_EOD_COORD", double, bridge::ProtocolType::INDI);
            double currentDec = GET_DEVICE_PROPERTY_AUTO(Telescope, "telescope1", "EQUATORIAL_EOD_COORD", double, bridge::ProtocolType::INDI);
            
            std::cout << "    Current coordinates: RA=" << currentRA << "h, Dec=" << currentDec << "°" << std::endl;
            
            // Set target coordinates
            SET_DEVICE_PROPERTY_AUTO(Telescope, "telescope1", "EQUATORIAL_EOD_COORD", 12.5, bridge::ProtocolType::INDI);
            std::cout << "    Target RA set to 12.5h" << std::endl;
            
            // Start slew
            std::cout << "    Starting slew to target..." << std::endl;
            INVOKE_DEVICE_METHOD_AUTO(Telescope, void, "telescope1", "slewToCoordinates", bridge::ProtocolType::INDI, 12.5, 45.0);
            
            // Check slewing status
            bool slewing = GET_DEVICE_PROPERTY_AUTO(Telescope, "telescope1", "TELESCOPE_MOTION_NS", bool, bridge::ProtocolType::INDI);
            std::cout << "    Slewing: " << (slewing ? "Yes" : "No") << std::endl;
            
        } catch (const std::exception& e) {
            std::cout << "    Error: " << e.what() << std::endl;
        }
    }
    
    void showStatistics() {
        std::cout << "\n6. Integration Statistics..." << std::endl;
        
        auto stats = manager_->getStatistics();
        
        std::cout << "  Total devices: " << stats.totalDevices << std::endl;
        std::cout << "  ASCOM-enabled devices: " << stats.ascomDevices << std::endl;
        std::cout << "  INDI-enabled devices: " << stats.indiDevices << std::endl;
        std::cout << "  Uptime: " << stats.uptime.count() << " ms" << std::endl;
        
        std::cout << "  Device types:" << std::endl;
        for (const auto& [type, count] : stats.deviceTypeCount) {
            std::cout << "    " << type << ": " << count << std::endl;
        }
        
        // Show registered devices
        auto deviceIds = manager_->getRegisteredDeviceIds();
        std::cout << "  Registered devices:" << std::endl;
        for (const auto& deviceId : deviceIds) {
            std::cout << "    " << deviceId << std::endl;
        }
    }
    
    void cleanup() {
        std::cout << "\n7. Cleaning up..." << std::endl;
        
        // Unregister devices
        UNREGISTER_DEVICE_AUTO("camera1");
        UNREGISTER_DEVICE_AUTO("telescope1");
        UNREGISTER_DEVICE_AUTO("focuser1");
        
        // Stop devices
        if (camera_) {
            camera_->stopDevice();
        }
        if (telescope_) {
            telescope_->stopDevice();
        }
        if (focuser_) {
            focuser_->stopDevice();
        }
        
        // Stop integration manager
        manager_->stop();
        
        std::cout << "  Cleanup completed" << std::endl;
    }
};

/**
 * @brief Main function demonstrating the automatic compatibility system
 */
int main() {
    try {
        CompatibilityExample example;
        example.runExample();
        return 0;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
}
